chains/solana/contracts/programs/ccip-router/src/instructions/v1/token_admin_registry.rs

```diff
@@ -1,77 +1,151 @@
+use crate::{
+    events::token_admin_registry as events,
+    token_context::{
+        OverridePendingTokenAdminRegistryByCCIPAdmin, OverridePendingTokenAdminRegistryByOwner,
+    },
+};
 use anchor_lang::prelude::*;
 use anchor_spl::associated_token::get_associated_token_address_with_program_id;
 use solana_program::{address_lookup_table::state::AddressLookupTable, log::sol_log};
 
 use super::pools::token_admin_registry_writable;
 
 use crate::{
-    seed, AcceptAdminRoleTokenAdminRegistry, AdministratorRegistered,
-    AdministratorTransferRequested, AdministratorTransferred, CcipRouterError,
-    ModifyTokenAdminRegistry, PoolSet, RegisterTokenAdminRegistryViaGetCCIPAdmin,
-    RegisterTokenAdminRegistryViaOwner, SetPoolTokenAdminRegistry,
+    seed,
+    token_context::{RegisterTokenAdminRegistryByCCIPAdmin, RegisterTokenAdminRegistryByOwner},
+    AcceptAdminRoleTokenAdminRegistry, CcipRouterError, ModifyTokenAdminRegistry,
+    SetPoolTokenAdminRegistry,
 };
 
 const MINIMUM_TOKEN_POOL_ACCOUNTS: usize = 9;
 
-pub fn register_token_admin_registry_via_get_ccip_admin(
-    ctx: Context<RegisterTokenAdminRegistryViaGetCCIPAdmin>,
-    mint: Pubkey, // should we validate that this is a real token program?
+// Proposes an administrator for the given token as pending administrator
+pub fn ccip_admin_propose_administrator(
+    ctx: Context<RegisterTokenAdminRegistryByCCIPAdmin>,
+    token_admin_registry_admin: Pubkey,
+) -> Result<()> {
+    let token_mint = ctx.accounts.mint.key().to_owned();
+    let token_admin_registry = &mut ctx.accounts.token_admin_registry;
+
+    init_with_pending_administrator(token_admin_registry, token_mint, token_admin_registry_admin)?;
+
+    Ok(())
+}
+
+// Overrides the pending administrator for the given token
+pub fn ccip_admin_override_pending_administrator(
+    ctx: Context<OverridePendingTokenAdminRegistryByCCIPAdmin>,
+    token_admin_registry_admin: Pubkey,
+) -> Result<()> {
+    let token_admin_registry = &mut ctx.accounts.token_admin_registry;
+    token_admin_registry.pending_administrator = token_admin_registry_admin;
+
+    Ok(())
+}
```

Comment:
when is this expected to be called?

it seems like this can be called any time which allows the ccip admin or token owner to override the pending admin at any time

my understanding is that once an `administrator` is set, then the token owner and ccip admin can no longer change anything

Reply:
so maybe there needs to be one more constraint added to the contexts?

Reply:
Oh, good catch! I forgot to add a check for `administrator == 0`. As this method is intended to be used when a missconfiguration happened. Will add a test for that!